### PR TITLE
Detailed Intl.DateTimeFormatOptions types

### LINKED ISSUE #35865: Missing specific types in DateTimeFormatOptions
**TypeScript Version:**  3.7.3

**Search Terms:** Intl.DateTimeFormat, DateTimeFormatOptions, date format types

**Code**

```ts
// A *self-contained* demonstration of the problem follows...
// Test this by running `tsc` on the command-line, rather than through another build tool such as Gulp,
(See below)
```
DateTimeFormatOptions interface missing specific types.
https://github.com/microsoft/TypeScript/blob/408b17649a1197a52f68fcb49b8c2f1eeac13668/src/lib/es5.d.ts#L4253-L4267

**Expected behavior:**
I would expect the types to match what's listed here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DateTimeFormat

```ts
  interface DateTimeFormatOptions {
        localeMatcher?: 'lookup' | 'best fit';
        weekday?: 'long' | 'short' | 'narrow';
        era?:  'long' | 'short' | 'narrow';
        year?: 'numeric' | '2-digit';
        month?: 'numeric' | '2-digit' | 'long' | 'short' | 'narrow';
        day?: 'numeric' | '2-digit';
        hour?: 'numeric' | '2-digit';
        minute?: 'numeric' | '2-digit';
        second?: 'numeric' | '2-digit';
        timeZoneName?: 'long' | 'short';
        formatMatcher?: 'basic' | 'best fit';
        hour12?: boolean;
        timeZone?: string; // this is more complicated than the others, not sure what I expect here
    }
```
**Actual behavior:**
```ts
 interface DateTimeFormatOptions {
        localeMatcher?: string;
        weekday?: string;
        era?: string;
        year?: string;
        month?: string;
        day?: string;
        hour?: string;
        minute?: string;
        second?: string;
        timeZoneName?: string;
        formatMatcher?: string;
        hour12?: boolean;
        timeZone?: string;
    }
```
**Playground Link:** 

**Related Issues:** 
Couldn't find any

### PR DESCRIPTION
**Pull request by johan**
Fixes #35865: Missing specific types in DateTimeFormatOptions.

This narrows the type info for `Intl.DateTimeFormatOptions` to legal inputs.

Existing code passing options satisfying the former (overly-lax) types in shipping typescript (and which browsers would thus throw a run-time `RangeError` exception for), will now be flagged as typescript compile-time errors, instead.

### CODE REVIEW

**Comment by johan** [comment]
FWIW: the failing tests here are present on master and probably every other branch too (covering whatever might be right or wrong about my implementation here). They look like this, if this is the first time you hear of them:

```
> ./node_modules/.bin/gulp runtests
[08:28:36] Using gulpfile ~/vc/TypeScript/gulpfile.js
[08:28:36] Starting 'runtests'...
[08:28:36] Starting 'generateLibs'...
[08:28:36] Starting 'buildScripts'...
[08:28:36] Finished 'generateLibs' after 194 ms
[08:28:37] Finished 'buildScripts' after 369 ms
[08:28:37] Starting 'generateDiagnostics'...
[08:28:37] Finished 'generateDiagnostics' after 579 μs
[08:28:37] Starting 'buildShims'...
[08:28:37] Starting 'buildDebugTools'...
[08:28:37] Finished 'buildShims' after 303 ms
[08:28:37] Finished 'buildDebugTools' after 303 ms
[08:28:37] Starting 'buildTsc'...
[08:28:37] Starting 'buildTests'...
[08:28:37] Starting 'buildTypescriptServicesOut'...
[08:28:37] Starting 'buildServerLibraryOut'...
src/harness/vfsUtil.ts:666:47 - error TS2345: Argument of type 'string' is not assignable to parameter of type '"ascii" | "utf8" | "utf-8" | "utf16le" | "ucs2" | "ucs-2" | "base64" | "latin1" | "binary" | "hex" | undefined'.

666             return encoding ? buffer.toString(encoding) : buffer;
                                                  ~~~~~~~~


Found 1 error.

[08:28:40] 'buildTsc' errored after 3.1 s
[08:28:40] Error: Process exited with code: 1
    at ChildProcess.<anonymous> (/Users/jhs/vc/TypeScript/scripts/build/utils.js:54:28)
    at ChildProcess.emit (events.js:311:20)
    at ChildProcess.EventEmitter.emit (domain.js:505:15)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:275:12)
[08:28:40] 'runtests' errored after 3.84 s
[08:28:40] The following tasks did not complete: buildTests, <series>, <series>, buildTypescriptServicesOut, buildServerLibraryOut
[08:28:40] Did you forget to signal async completion?
```

**Comment by johan** [comment]
Ah, that was addressed by @Kingwl in #38525 – thank you! Rebased this PR on top of that fix.

File: src/lib/es5.d.ts [hunk 1]
```diff
@@ -4277,20 +4277,34 @@ declare namespace Intl {
         supportedLocalesOf(locales: string | string[], options?: NumberFormatOptions): string[];
     };
 
+    type DateFormatNumericVerbosity = "numeric" | "2-digit";
+    type DateFormatNameVerbosity = "long" | "short" | "narrow";
```
Comment (sandersn) [inline src/lib/es5.d.ts:4281]: Please inline the type aliases. They're needlessly polluting the global namespace. And if the alias name makes it into quickinfo, it just obscures what the legal values are.

Comment (johan) [reply src/lib/es5.d.ts]: Thanks, and addressed!

**Comment by johan** [comment]
@sandersn, are you taking the opportunity with the 4.0 major release to land PRs like this that stricten overly-loose types?

**Comment by sandersn** [comment]
For this PR, I'm waiting for the originating issue #35865 to be discussed and accepted. That happens during regular backlog triage meetings, and is sometimes followed by a design meeting discussion with the whole team.

In general, Typescript doesn't follow semver for a couple of reasons. The first is historical -- we started with the old-school point-release version numbers on a regular release schedule. The second is that is our product is all interface -- *every* release breaks things, and it's not easy to predict which things will break in a large code base, so there's little value in major-vs-minor releases.

**Comment by sandersn** [comment]
This PR has been blocked for quite a while, so I'm going to close it to keep the number of open PRs manageable. We can re-open it if #35865 gets accepted.

**Pull request closed without merging**